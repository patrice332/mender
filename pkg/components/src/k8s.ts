import { type ApiObjectMetadata, App, Chart } from "cdk8s";
import * as kplus from "cdk8s-plus-32";
import { Construct } from "constructs";

export { ImagePullPolicy } from "cdk8s-plus-32";

export class K8s {
  public readonly underlying: App;
  private readonly projectNames: string[];
  private readonly projects: Record<string, Project>;

  constructor() {
    this.underlying = new App();
    this.projectNames = [];
    this.projects = {};
  }

  public createProject(name: string): Project {
    if (name in this.projects) {
      throw new Error(`Project with name (${name}) already exists`);
    }
    const project = new Project(this.underlying, name);
    this.projectNames.push(name);
    this.projects[name] = project;
    return project;
  }

  public synth = (): unknown[] =>
    this.projectNames.flatMap((name) => {
      const project = this.projects[name];
      if (!project) {
        throw new Error(`Internal Error: Project named (${name}) is in project list, but cannot be found`);
      }
      return project.underlying.toJson();
    });
}

export class Project {
  public readonly underlying: Chart;
  public readonly name: string;

  constructor(app: App, name: string) {
    this.underlying = new Chart(app, name);
    this.name = name;
  }

  public createNamespace(
    name: string,
    props: Omit<kplus.NamespaceProps, "metadata"> & { metadata?: Omit<ApiObjectMetadata, "name"> },
  ): Namespace {
    return new Namespace(this.underlying, name, props);
  }
}

export class Namespace extends Construct {
  private readonly chart: Chart;
  public readonly underlying: kplus.Namespace;
  public readonly name: string;

  constructor(
    scope: Chart,
    name: string,
    props: Omit<kplus.NamespaceProps, "metadata"> & { metadata?: Omit<ApiObjectMetadata, "name"> },
  ) {
    super(scope, name);
    this.chart = scope;
    const localProps: kplus.NamespaceProps = {
      ...props,
      metadata: {
        ...props.metadata,
        name,
      },
    };
    this.underlying = new kplus.Namespace(scope, `${scope.node.path.replaceAll("/", "-")}-namespace`, localProps);
    this.name = name;
  }

  public createDeployment(
    name: string,
    props: Omit<kplus.DeploymentProps, "metadata"> & { metadata?: Omit<ApiObjectMetadata, "name" | "namespace"> },
  ): Deployment {
    return new Deployment(this.chart, this, name, props);
  }
}

export class Deployment extends Construct {
  public readonly underlying: kplus.Deployment;
  public readonly name: string;

  constructor(
    scope: Chart,
    namespace: Namespace,
    name: string,
    props: Omit<kplus.DeploymentProps, "metadata"> & { metadata?: Omit<ApiObjectMetadata, "name" | "namespace"> },
  ) {
    super(scope, name);
    const localProps: kplus.DeploymentProps = {
      ...props,
      metadata: {
        ...props.metadata,
        name,
        namespace: namespace.name,
      },
    };
    this.underlying = new kplus.Deployment(scope, `${namespace.node.path.replaceAll("/", "-")}-deployment`, localProps);
    this.name = name;
  }
}
