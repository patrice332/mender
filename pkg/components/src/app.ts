import { RootConstruct } from "constructs";
import { K8s } from "./k8s";

export class App extends RootConstruct {
  public readonly k8s: K8s;
  constructor() {
    super();
    this.k8s = new K8s();
  }
}
