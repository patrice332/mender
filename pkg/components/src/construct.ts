import { Construct as BaseConstruct } from "constructs";
import { App } from "./app";
import type { K8s } from "./k8s";

export class Construct extends BaseConstruct {
  // biome-ignore lint/complexity/noUselessConstructor: Adds extra constraints on scope
  constructor(scope: Construct | App, id: string) {
    super(scope, id);
  }

  get k8s(): K8s {
    const root = this.node.root;
    if (!(root instanceof App)) {
      throw new Error("Construct is not a child of a mender app");
    }
    return root.k8s;
  }
}
