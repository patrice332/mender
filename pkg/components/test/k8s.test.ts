import { describe, expect, it } from "vitest";
import { addDeployment, addManifest, ImagePullPolicy, K8s } from "../src/k8s";

describe("k8s", () => {
  it("can create a deployment", () => {
    const kplus = new K8s();
    const project = kplus.createProject("test");
    const ns = project.createNamespace("test-ns", {
      metadata: {
        labels: {
          "test.label": "value",
        },
        annotations: {
          "test.annotate": "other_value",
        },
      },
    });
    addDeployment(ns, "test-deploy", {
      metadata: {
        labels: {
          "deploy-label": "foo",
        },
        annotations: {
          "deploy-annotate": "bar",
        },
      },
      containers: [
        {
          name: "test-container",
          image: "test-image:sometag",
          imagePullPolicy: ImagePullPolicy.IF_NOT_PRESENT,
        },
      ],
    });

    expect(kplus.synth()).toMatchSnapshot();
  });

  it("can create raw manifests", () => {
    const kplus = new K8s();

    const project = kplus.createProject("test");
    const ns = project.createNamespace("test-ns", {
      metadata: {
        labels: {
          "test.label": "value",
        },
        annotations: {
          "test.annotate": "other_value",
        },
      },
    });

    addManifest(ns, "foo-obj", {
      apiVersion: "v1",
      kind: "MyCustomCRD",
    });

    expect(kplus.synth()).toMatchSnapshot();
  });
});
