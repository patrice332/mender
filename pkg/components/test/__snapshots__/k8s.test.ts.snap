// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`k8s > can create a deployment 1`] = `
[
  {
    "apiVersion": "v1",
    "kind": "Namespace",
    "metadata": {
      "annotations": {
        "test.annotate": "other_value",
      },
      "labels": {
        "test.label": "value",
      },
      "name": "test-ns",
    },
    "spec": {},
  },
  {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": {
      "annotations": {
        "deploy-annotate": "bar",
      },
      "labels": {
        "deploy-label": "foo",
      },
      "name": "test-deploy",
      "namespace": "test-ns",
    },
    "spec": {
      "minReadySeconds": 0,
      "progressDeadlineSeconds": 600,
      "replicas": 2,
      "revisionHistoryLimit": 10,
      "selector": {
        "matchLabels": {
          "cdk8s.io/metadata.addr": "test-test-test-ns-deployment-c8940b74",
        },
      },
      "strategy": {
        "rollingUpdate": {
          "maxSurge": "25%",
          "maxUnavailable": "25%",
        },
        "type": "RollingUpdate",
      },
      "template": {
        "metadata": {
          "labels": {
            "cdk8s.io/metadata.addr": "test-test-test-ns-deployment-c8940b74",
          },
        },
        "spec": {
          "automountServiceAccountToken": false,
          "containers": [
            {
              "image": "test-image:sometag",
              "imagePullPolicy": "IfNotPresent",
              "name": "test-container",
              "resources": {
                "limits": {
                  "cpu": "1500m",
                  "memory": "2048Mi",
                },
                "requests": {
                  "cpu": "1000m",
                  "memory": "512Mi",
                },
              },
              "securityContext": {
                "allowPrivilegeEscalation": false,
                "privileged": false,
                "readOnlyRootFilesystem": true,
                "runAsNonRoot": true,
              },
            },
          ],
          "dnsPolicy": "ClusterFirst",
          "hostNetwork": false,
          "restartPolicy": "Always",
          "securityContext": {
            "fsGroupChangePolicy": "Always",
            "runAsNonRoot": true,
          },
          "setHostnameAsFQDN": false,
          "shareProcessNamespace": false,
          "terminationGracePeriodSeconds": 30,
        },
      },
    },
  },
]
`;
